scico/linop/xray/_xray.py

```diff
@@ -193,3 +195,128 @@ def _calc_weights(x0, dx, nx, angle, y0):
     weights = jnp.minimum(distance_to_next, width) / width
 
     return inds, weights
+
+
+class Parallel3dProjector:
+    """General-purpose, 3D, parallel ray X-ray projector."""
+
+    def __init__(
+        self,
+        input_shape: Shape,
+        matrices: ArrayLike,
+        det_shape: Shape,
+    ):
+        r"""
+        Args:
+            input_shape: Shape of input image.
+            matrices: (num_angles, 2, 4) array of homogeneous projection matrices.
+            det_shape: Shape of detector.
+        """
+
+        self.input_shape = input_shape
+        self.matrices = matrices
+        self.det_shape = det_shape
+        self.output_shape = (len(matrices), *det_shape)
+
+    def project(self, im):
+        """Compute X-ray projection."""
+        return Parallel3dProjector._project(im, self.matrices, self.det_shape)
+
+    @staticmethod
+    def _project(im: ArrayLike, matrices: ArrayLike, det_shape: Shape) -> ArrayLike:
+        r"""
+        Args:
+            im: Input image.
+            matrix: (num_views, 2, 4) array of homogeneous projection matrices.
+            det_shape: Shape of detector.
+        """
+        MAX_SLICE_LEN = 10
+        slice_offsets = list(range(0, im.shape[0], MAX_SLICE_LEN))
+
+        num_views = len(matrices)
+        proj = jnp.zeros((num_views,) + det_shape, dtype=im.dtype)
+        for view_ind, matrix in enumerate(matrices):
+            for slice_offset in slice_offsets:
```

Comment:
Would it be possible to replace these Python loops with a `jax.lax` control structure? This code is currently not jit-able.

Reply:
I'm not sure. I had a hard time determining which primitives would loop the way I wanted (i.e., not parallelize too much and run out of memory for large projections on modest GPUs).  [`jax.map`](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.map.html) might be a candidate because it has a `batch_size` option that controls how much happens in parallel. I don't think this solution is perfect, but it will be nontrivial to improve.